gov/src/lib.rs

```diff
@@ -0,0 +1,248 @@
+            rt.block_on(run_propose(file, governance))?;
+        }
+        Command::Vote => {
+            let id = options.proposal_id.unwrap();
+            rt.block_on(run_vote(id, governance))?;
+        }
+    }
+
+    Ok(())
+}
+
+async fn run_propose<M>(file: OsString, governance: Governance<M>) -> anyhow::Result<()>
+where
+    M: Middleware + 'static,
+    crate::governance::Error<M>: From<<M as Middleware>::Error>,
+{
+    let spinner = term::spinner(&format!(
+        "Reading proposal file {}",
+        term::format::highlight(&format!("{:?}", file))
+    ));
+
+    let content = read_to_string(file).expect("Couldn't read proposal file.");
+    let cutoff = content
+        .lines()
+        .position(|l| l == "## ACTIONS ##")
+        .ok_or_else(|| anyhow!("Proposal is missing `### ACTIONS ###` section."))?;
+
+    let mut targets: Vec<Address> = Vec::new();
+    let mut values: Vec<U256> = Vec::new();
+    let mut signatures: Vec<String> = Vec::new();
+    let mut calldatas: Vec<Vec<u8>> = Vec::new();
+
+    for l in content.lines().skip(cutoff + 1) {
+        if l == "```" || l.is_empty() {
+            continue;
+        }
+        let mut tokens = l.split(' ');
+
+        let address = tokens.next().unwrap();
+        let value = tokens.next().unwrap();
+        let sig = tokens.next().map(quoteless_string).unwrap();
+
+        let function = AbiParser::default().parse_function(&sig).unwrap();
+
+        let args: Vec<String> = tokens.map(quoteless_string).collect();
+        let args: Vec<&str> = args.iter().map(|t| t.as_str()).collect();
+        let params: Vec<_> = function
+            .inputs
+            .iter()
+            .map(|param| param.kind.clone())
+            .zip(args)
+            .collect();
+
+        let tokens: anyhow::Result<Vec<Token>> = params
+            .iter()
+            .map(|&(ref param, value)| LenientTokenizer::tokenize(param, value))
+            .collect::<Result<_, _>>()
+            .map_err(From::from);
+        let tokens = tokens?;
+
+        let calldata = if !tokens.is_empty() {
+            let mut calldata = function.encode_input(&tokens)?;
+            calldata.drain(..4);
+            calldata
+        } else {
+            Vec::new()
+        };
+
+        targets.push(Address::from_str(address).unwrap());
+        values.push(U256::from_dec_str(value).unwrap());
+        signatures.push(sig.to_string());
+        calldatas.push(calldata.clone());
+    }
+    spinner.finish();
+
+    let spinner = term::spinner("Sending proposal...");
+    let call = governance.propose(targets, values, signatures, calldatas, content)?;
+    ethereum::transaction(call).await?;
+    spinner.finish();
+
+    Ok(())
+}
+
+async fn run_vote<M>(id: U256, governance: Governance<M>) -> anyhow::Result<()>
+where
+    M: Middleware + 'static,
+    crate::governance::Error<M>: From<<M as Middleware>::Error>,
+{
+    let proposal = governance.get_proposal(id).await?.unwrap();
```

Comment:
We can do `.await??` here maybe to return a nicer error in case.

Reply:
Actually we don't need `Option` here.